# Project - Data Cleaning

## Introduction
In this lab, we'll make use of everything we've learned about pandas, data cleaning, and Exploratory Data Analysis. In order to complete this lab, you'll have to make import, clean, combine, reshape, and visualize data to answer questions provided, as well as your own questions!

## Objectives
You will be able to:
* Show mastery of the content covered in this section

## The Dataset
In this lab, we'll work with the comprehensive [Super Heroes Dataset](https://www.kaggle.com/claudiodavi/superhero-set/data), which can be found on Kaggle!

## Goals
* Use all available pandas knowledge to clean the dataset and deal with null values
* Use Queries and aggregations to group the data into interesting subsets as needed
* Use descriptive statistics and data visualization to find answers to questions we may have about the data. 

## Getting Started

In the cell below:

* Import and alias pandas as `pd`
* Import and alias numpy as `np`
* Import and alias seaborn as `sns`
* Import and alias matplotlib.pyplot as `plt`
* Set matplotlib visualizations to display inline in the notebook

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

For this lab, our dataset is split among two different sources--`heroes_information.csv` and `super_hero_powers.csv`.

Use pandas to read in each file and store them in DataFrames in the appropriate variables below. Then, display the head of each to ensure that everything loaded correctly.  

In [66]:
heroes_df = pd.read_csv('heroes_information.csv')
powers_df = pd.read_csv('super_hero_powers.csv')
heroes_df.head()

Unnamed: 0         name Gender Eye color               Race Hair color  \
0           0       A-Bomb   Male    yellow              Human    No Hair   
1           1   Abe Sapien   Male      blue      Icthyo Sapien    No Hair   
2           2     Abin Sur   Male      blue            Ungaran    No Hair   
3           3  Abomination   Male     green  Human / Radiation    No Hair   
4           4      Abraxas   Male      blue      Cosmic Entity      Black   

   Height          Publisher Skin color Alignment  Weight  
0   203.0      Marvel Comics          -      good   441.0  
1   191.0  Dark Horse Comics       blue      good    65.0  
2   185.0          DC Comics        red      good    90.0  
3   203.0      Marvel Comics          -       bad   441.0  
4   -99.0      Marvel Comics          -       bad   -99.0

In [67]:
powers_df.head()

hero_names  Agility  Accelerated Healing  Lantern Power Ring  \
0      3-D Man     True                False               False   
1       A-Bomb    False                 True               False   
2   Abe Sapien     True                 True               False   
3     Abin Sur    False                False                True   
4  Abomination    False                 True               False   

   Dimensional Awareness  Cold Resistance  Durability  Stealth  \
0                  False            False       False    False   
1                  False            False        True    False   
2                  False             True        True    False   
3                  False            False       False    False   
4                  False            False       False    False   

   Energy Absorption  Flight     ...      Web Creation  Reality Warping  \
0              False   False     ...             False            False   
1              False   False     ...             False            False   
2              False   False     ...             False            False   
3              False   False     ...             False            False   
4              False   False     ...             False            False   

   Odin Force  Symbiote Costume  Speed Force  Phoenix Force  \
0       False             False        False          False   
1       False             False        False          False   
2       False             False        False          False   
3       False             False        False          False   
4       False             False        False          False   

   Molecular Dissipation  Vision - Cryo  Omnipresent  Omniscient  
0                  False          False        False       False  
1                  False          False        False       False  
2                  False          False        False       False  
3                  False          False        False       False  
4                  False          False        False       False  

[5 rows x 168 columns]

It looks as if the heroes information dataset contained an index column.  We did not specify that this dataset contained an index column, because we hadn't seen it yet. Pandas does not know how to tell apart an index column from any other data, so it stored it with the column name `Unnamed: 0`.  

Our DataFrame provided row indices by default, so this column is not needed.  Drop it from the DataFrame in place in the cell below, and then display the head of `heroes_df` to ensure that it worked properly. 

In [68]:
heroes_df.drop(['Unnamed: 0'], axis=1, inplace=True)

In [69]:
heroes_df.head()

name Gender Eye color               Race Hair color  Height  \
0       A-Bomb   Male    yellow              Human    No Hair   203.0   
1   Abe Sapien   Male      blue      Icthyo Sapien    No Hair   191.0   
2     Abin Sur   Male      blue            Ungaran    No Hair   185.0   
3  Abomination   Male     green  Human / Radiation    No Hair   203.0   
4      Abraxas   Male      blue      Cosmic Entity      Black   -99.0   

           Publisher Skin color Alignment  Weight  
0      Marvel Comics          -      good   441.0  
1  Dark Horse Comics       blue      good    65.0  
2          DC Comics        red      good    90.0  
3      Marvel Comics          -       bad   441.0  
4      Marvel Comics          -       bad   -99.0

## Familiarize Yourself With the Dataset

The first step in our Exploratory Data Analysis will be to get familiar with the data.  This step includes:

* Understanding the dimensionality of your dataset
* Investigating what type of data it contains, and the data types used to store it
* Discovering how missing values are encoded, and how many there are
* Getting a feel for what information it does and doesnt contain

In the cell below, get the descriptive statistics of each DataFrame.  

In [70]:
heroes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 734 entries, 0 to 733
Data columns (total 10 columns):
name          734 non-null object
Gender        734 non-null object
Eye color     734 non-null object
Race          734 non-null object
Hair color    734 non-null object
Height        734 non-null float64
Publisher     719 non-null object
Skin color    734 non-null object
Alignment     734 non-null object
Weight        732 non-null float64
dtypes: float64(2), object(8)
memory usage: 57.4+ KB


In [71]:
heroes_df.describe()

Height      Weight
count  734.000000  732.000000
mean   102.254087   43.855191
std    139.624543  130.823733
min    -99.000000  -99.000000
25%    -99.000000  -99.000000
50%    175.000000   62.000000
75%    185.000000   90.000000
max    975.000000  900.000000

In [72]:
heroes_df['Gender'].value_counts()

Male      505
Female    200
-          29
Name: Gender, dtype: int64

In [73]:
for c in heroes_df.columns:
    print(f'-----------Column: {c}-----------')
    print(heroes_df[c].value_counts().head(10))

-----------Column: name-----------
Spider-Man      3
Goliath         3
Atlas           2
Blue Beetle     2
Speedy          2
Angel           2
Black Canary    2
Blizzard        2
Batman          2
Atom            2
Name: name, dtype: int64
-----------Column: Gender-----------
Male      505
Female    200
-          29
Name: Gender, dtype: int64
-----------Column: Eye color-----------
blue      225
-         172
brown     126
green      73
red        46
black      23
yellow     19
white      17
hazel       6
grey        6
Name: Eye color, dtype: int64
-----------Column: Race-----------
-                    304
Human                208
Mutant                63
God / Eternal         14
Human / Radiation     11
Cyborg                11
Android                9
Symbiote               9
Kryptonian             7
Alien                  7
Name: Race, dtype: int64
-----------Column: Hair color-----------
-                   172
Black               158
Blond                99
Brown                

In [74]:
heroes_df.isna().sum()

name           0
Gender         0
Eye color      0
Race           0
Hair color     0
Height         0
Publisher     15
Skin color     0
Alignment      0
Weight         2
dtype: int64

## Dealing with Null Values

Starting in the cell below, detect and deal with any null values in either data frame.  Then, explain your methodology for detecting and dealing with outliers in the markdown section below.  Be sure to explain your strategy for dealing with null values in numeric columns, as well as your strategy for dealing with null values in non-numeric columns.  

Note that if you need to add more cells to write code in, you can do this by:

**1.** Highlighting a cell and then pressing `ESC` to enter command mode.  
**2.** Press `A` to add a cell above the highlighted cell, or `B` to add a cell below the highlighted cell. 

Describe your strategy below this line:
____________________________________________________________________________________________________________________________




**In heroes_df:**


heroes_df.df['Publisher'] contains 7 null values as NaN --> drop rows


heroes_df.df['Weight'] contains 2 null values as NaN --> drop rows (rows with null values dropped with dropping either 'Race' or 'Skin colour' columns)


heroes_df.df['Gender'] contains 29 values as '-' --> keep as 'Unknown'


heroes_df.df['Eye color'] contains 172 values as '-' --> keep as 'Unknown'


heroes_df.df['Race'] contains 304 values as '-' --> drop column


heroes_df.df['Hair color'] contains 172 values as '-' --> keep as 'Unknown'


heroes_df.df['Skin color'] contains 662 values as '-' --> drop column


heroes_df.df['Alignment'] contains 7 values as '-' --> drop rows

In [75]:
heroes_df.drop(['Race'], axis=1, inplace=True)
heroes_df.drop(['Skin color'], axis=1, inplace=True)

In [76]:
heroes_df.dropna(subset=['Publisher'], inplace=True)

In [88]:
heroes_df = heroes_df[heroes_df['Alignment'] != '-']

**In powers_df:**

No null values.

In [96]:
powers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Columns: 168 entries, hero_names to Omniscient
dtypes: bool(167), object(1)
memory usage: 114.1+ KB


In [99]:
powers_df.isna().sum().sum()

0

In [100]:
for c in powers_df.columns:
    print(f'-----------Column: {c}-----------')
    print(powers_df[c].value_counts().head(5))

-----------Column: hero_names-----------
Lord Voldemort     1
Plantman           1
Cottonmouth        1
Gorilla Grodd      1
Kathryn Janeway    1
Name: hero_names, dtype: int64
-----------Column: Agility-----------
False    425
True     242
Name: Agility, dtype: int64
-----------Column: Accelerated Healing-----------
False    489
True     178
Name: Accelerated Healing, dtype: int64
-----------Column: Lantern Power Ring-----------
False    656
True      11
Name: Lantern Power Ring, dtype: int64
-----------Column: Dimensional Awareness-----------
False    642
True      25
Name: Dimensional Awareness, dtype: int64
-----------Column: Cold Resistance-----------
False    620
True      47
Name: Cold Resistance, dtype: int64
-----------Column: Durability-----------
False    410
True     257
Name: Durability, dtype: int64
-----------Column: Stealth-----------
False    541
True     126
Name: Stealth, dtype: int64
-----------Column: Energy Absorption-----------
False    590
True      77
Name: Ene

False    644
True      23
Name: Super Breath, dtype: int64
-----------Column: Wallcrawling-----------
False    633
True      34
Name: Wallcrawling, dtype: int64
-----------Column: Vision - Night-----------
False    633
True      34
Name: Vision - Night, dtype: int64
-----------Column: Vision - Infrared-----------
False    645
True      22
Name: Vision - Infrared, dtype: int64
-----------Column: Grim Reaping-----------
False    664
True       3
Name: Grim Reaping, dtype: int64
-----------Column: Matter Absorption-----------
False    656
True      11
Name: Matter Absorption, dtype: int64
-----------Column: The Force-----------
False    661
True       6
Name: The Force, dtype: int64
-----------Column: Resurrection-----------
False    652
True      15
Name: Resurrection, dtype: int64
-----------Column: Terrakinesis-----------
False    665
True       2
Name: Terrakinesis, dtype: int64
-----------Column: Vision - Heat-----------
False    648
True      19
Name: Vision - Heat, dtype: int64
---

## Joining, Grouping, and Aggregating

In the cell below, join the two DataFrames.  Think about which sort of join you should use, as well as which columns you should join on.  Rename columns and manipulate as needed.  

**_HINT:_** If the join throws an error message, consider settin the the column you want to join on as the index for each DataFrame.  

In [120]:
heroes_df.set_index('name', inplace=True)

In [122]:
powers_df.set_index('hero_names', inplace=True)

In [121]:
heroes_df.head()

Gender Eye color Hair color  Height          Publisher Alignment  \
name                                                                           
A-Bomb        Male    yellow    No Hair   203.0      Marvel Comics      good   
Abe Sapien    Male      blue    No Hair   191.0  Dark Horse Comics      good   
Abin Sur      Male      blue    No Hair   185.0          DC Comics      good   
Abomination   Male     green    No Hair   203.0      Marvel Comics       bad   
Abraxas       Male      blue      Black   -99.0      Marvel Comics       bad   

             Weight  
name                 
A-Bomb        441.0  
Abe Sapien     65.0  
Abin Sur       90.0  
Abomination   441.0  
Abraxas       -99.0

In [123]:
powers_df.head()

Agility  Accelerated Healing  Lantern Power Ring  \
hero_names                                                      
3-D Man         True                False               False   
A-Bomb         False                 True               False   
Abe Sapien      True                 True               False   
Abin Sur       False                False                True   
Abomination    False                 True               False   

             Dimensional Awareness  Cold Resistance  Durability  Stealth  \
hero_names                                                                 
3-D Man                      False            False       False    False   
A-Bomb                       False            False        True    False   
Abe Sapien                   False             True        True    False   
Abin Sur                     False            False       False    False   
Abomination                  False            False       False    False   

             Energy Absorption  Flight  Danger Sense     ...      \
hero_names                                               ...       
3-D Man                  False   False         False     ...       
A-Bomb                   False   False         False     ...       
Abe Sapien               False   False         False     ...       
Abin Sur                 False   False         False     ...       
Abomination              False   False         False     ...       

             Web Creation  Reality Warping  Odin Force  Symbiote Costume  \
hero_names                                                                 
3-D Man             False            False       False             False   
A-Bomb              False            False       False             False   
Abe Sapien          False            False       False             False   
Abin Sur            False            False       False             False   
Abomination         False            False       False             False   

             Speed Force  Phoenix Force  Molecular Dissipation  Vision - Cryo  \
hero_names                                                                      
3-D Man            False          False                  False          False   
A-Bomb             False          False                  False          False   
Abe Sapien         False          False                  False          False   
Abin Sur           False          False                  False          False   
Abomination        False          False                  False          False   

             Omnipresent  Omniscient  
hero_names                            
3-D Man            False       False  
A-Bomb             False       False  
Abe Sapien         False       False  
Abin Sur           False       False  
Abomination        False       False  

[5 rows x 167 columns]

In [124]:
#heroes_df.name = powers_df.hero_name
#left join
joined_df = heroes_df.join(powers_df, how='left', on='name')

In [125]:
joined_df.head()

Gender Eye color Hair color  Height          Publisher Alignment  \
name                                                                           
A-Bomb        Male    yellow    No Hair   203.0      Marvel Comics      good   
Abe Sapien    Male      blue    No Hair   191.0  Dark Horse Comics      good   
Abin Sur      Male      blue    No Hair   185.0          DC Comics      good   
Abomination   Male     green    No Hair   203.0      Marvel Comics       bad   
Abraxas       Male      blue      Black   -99.0      Marvel Comics       bad   

             Weight Agility Accelerated Healing Lantern Power Ring    ...      \
name                                                                  ...       
A-Bomb        441.0   False                True              False    ...       
Abe Sapien     65.0    True                True              False    ...       
Abin Sur       90.0   False               False               True    ...       
Abomination   441.0   False                True              False    ...       
Abraxas       -99.0   False               False              False    ...       

            Web Creation Reality Warping Odin Force Symbiote Costume  \
name                                                                   
A-Bomb             False           False      False            False   
Abe Sapien         False           False      False            False   
Abin Sur           False           False      False            False   
Abomination        False           False      False            False   
Abraxas            False           False      False            False   

            Speed Force Phoenix Force Molecular Dissipation Vision - Cryo  \
name                                                                        
A-Bomb            False         False                 False         False   
Abe Sapien        False         False                 False         False   
Abin Sur          False         False                 False         False   
Abomination       False         False                 False         False   
Abraxas           False         False                 False         False   

            Omnipresent Omniscient  
name                                
A-Bomb            False      False  
Abe Sapien        False      False  
Abin Sur          False      False  
Abomination       False      False  
Abraxas           False      False  

[5 rows x 174 columns]

In [126]:
len(joined_df)

713

In the cell below, subset male and female heroes into different dataframes.  Create a scatterplot of the height and weight of each hero, with weight as the y-axis.  Plot both the male and female heroes subset into each dataframe, and make the color for each point in the scatterplot correspond to the gender of the superhero.

## Some Initial Investigation

Next, slice the DataFrame as needed and visualize the distribution of heights and weights by gender.  You should have 4 total plots.  

In the cell below:

* Slice the DataFrame into separate DataFrames by gender
* Complete the `show_distplot` function.  This helper function should take in a DataFrame, a string containing the gender we want to visualize, and and the column name we want to visualize by gender. The function should display a distplot visualization from seaborn of the column/gender combination.  

Hint: Don't forget to check the [seaborn documentation for distplot](https://seaborn.pydata.org/generated/seaborn.distplot.html) if you have questions about how to use it correctly! 

In [127]:
male_heroes_df = None
female_heroes_df = None
# use loc?

In [ ]:
def show_distplot(dataframe, gender, column_name):
    pass

In [ ]:
# Male Height


In [ ]:
# Male Weight


In [ ]:
# Female Height


In [ ]:
# Female Weight


Discuss your findings from the plots above, with respect to the distibution of height and weight by gender.  Your explanation should include discussion of any relevant summary statistics, including mean, median, mode, and the overall shape of each distribution.  

Wite your answer below this line:
____________________________________________________________________________________________________________________________



### Sample Question: Most Common Powers

The rest of this notebook will be left to you to investigate the dataset by formulating your own questions, and then seeking answers using pandas and numpy.  Every answer should include some sort of visualization, when appropriate. Before moving on to formulating your own questions, use the dataset to answer the following questions about superhero powers:

* What are the 5 most common powers overall?
* What are the 5 most common powers in the Marvel Universe?
* What are the 5 most common powers in the DC Universe?

Analyze the results you found above to answer the following question:

How do the top 5 powers in the Marvel and DC universes compare?  Are they similar, or are there significant differences? How do they compare to the overall trends in the entire Superheroes dataset?

Wite your answer below this line:
____________________________________________________________________________________________________________________________


### Your Own Investigation

For the remainder of this lab, you'll be focusing on coming up with and answering your own question, just like we did above.  Your question should not be overly simple, and should require both descriptive statistics and data visualization to answer.  In case you're unsure of what questions to ask, some sample questions have been provided below.

Pick one of the following questions to investigate and answer, or come up with one of your own!

* Which powers have the highest chance of co-occuring in a hero (e.g. super strength and flight), and does this differ by gender?
* Is there a relationship between a hero's height and weight and their powerset?
* What is the distribution of skin colors amongst alien heroes?

Explain your question below this line:
____________________________________________________________________________________________________________________________



Some sample cells have been provided to give you room to work. If you need to create more cells, you can do this easily by:

1. Highlighting a cell and then pressing `esc` to enter command mode.
1. Pressing `b` to add a cell below the currently highlighted cell, or `a` to add one above it.  

Be sure to include thoughtful, well-labeled visualizations to back up your analysis!

## Summary

In this lab, we demonstrated our mastery of:
* Using all of our Pandas knowledge to date to clean the dataset and deal with null values
* Using Queries and aggregations to group the data into interesting subsets as needed
* Using descriptive statistics and data visualization to find answers to questions we may have about the data